src/core/common/settings.cpp

```diff
@@ -0,0 +1,508 @@
+        {
+            uint32_t address = sBase + kSettingsFlagSize;
+            struct stageDeleteSettingsBlock *stageDeleteBlock = &sStageDeleteSettingsBlock[stageDeleteIndex++];
+            struct settingsBlock block;
+
+            while (address < (sBase + sSettingsSize))
+            {
+                otPlatFlashRead(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+
+                if (!(block.flag & kBlockAddFlag) && (block.flag & kBlockDeleteFlag) &&
+                    (block.key == stageDeleteBlock->key) &&
+                    (block.index == static_cast<uint8_t>(stageDeleteBlock->index) || stageDeleteBlock->index == -1))
+                {
+                    block.flag &= (~kBlockDeleteFlag);
+                    otPlatFlashWrite(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+                    sSwapDone = false;
+                    VerifyOrExit(stageDeleteBlock->index == -1, ;);
+                }
+
+                address += (block.length + sizeof(struct settingsBlock));
+            }
+        }
+    }
+
+exit:
+    initCommitChanges(false);
+    return error;
+}
+
+ThreadError otPlatSettingsAbandonChange(void)
+{
+    ThreadError error = kThreadError_None;
+
+    VerifyOrExit(sCommitLock == true, error = kThreadError_InvalidState);
+    initCommitChanges(false);
+
+exit:
+    return error;
+}
+
+ThreadError otPlatSettingsGet(uint16_t aKey, int aIndex, uint8_t *aValue, int *aValueLength)
+{
+    ThreadError error = kThreadError_NotFound;
+    uint32_t address = sBase + kSettingsFlagSize;
+
+    while (address < (sBase + sSettingsSize))
+    {
+        struct settingsBlock block;
+
+        otPlatFlashRead(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+
+        if (block.key == aKey && block.index == static_cast<uint8_t>(aIndex) &&
+            (!(block.flag & kBlockAddFlag) && (block.flag & kBlockDeleteFlag)))
+        {
+            error = kThreadError_None;
+
+            if (aValueLength)
+            {
+                *aValueLength = block.length;
+            }
+
+            if (aValue)
+            {
+                VerifyOrExit(aValueLength, error = kThreadError_InvalidArgs);
+                otPlatFlashRead(address + sizeof(struct settingsBlock), aValue, block.length);
+            }
+
+            break;
+        }
+
+        address += (block.length + sizeof(struct settingsBlock));
+    }
+
+exit:
+    return error;
+}
+
+ThreadError otPlatSettingsSet(uint16_t aKey, const uint8_t *aValue, int aValueLength)
+{
+    ThreadError error = kThreadError_None;
+    uint32_t address = sBase + kSettingsFlagSize;
+
+    while (address < (sBase + sSettingsSize))
+    {
+        struct settingsBlock block;
+
+        otPlatFlashRead(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+
+        if ((!(block.flag & kBlockAddFlag) && block.flag & kBlockDeleteFlag) && (block.key == aKey))
+        {
+            SuccessOrExit(error = otPlatSettingsDelete(aKey, -1));
```

Comment:
I have changed the logic here.
otPlatSettingsSet is otPlatSettingsAdd with index 0.

The block entry is valid only when
- kSettingsAddComplete flag is not set
- kSettingsDelete flag is set
- there is not block entry containing the same key and <= index after this block entry